test/grouping.jl

```diff
@@ -22,4 +23,8 @@ d = DataFrame(n = 1:20, x = [3, 3, 3, 3, 1, 1, 1, 2, 1, 1, 2, 1, 1, 2, 2, 2, 3,
 g = groupby(d, :x, sort=true)
 @test @based_on(g, nsum = sum(:n))[:nsum] == [99, 84, 27]
 
+d = DataFrame(a = [1,1,1,2,2], b = [1,2,3,missing,missing])
+g = groupby(d, :a)
+@test isequal(@transform(g, t = mean(:b))[:t], [2.0, 2.0, 2.0, missing, missing])
```

Comment:
Can you also test what happens when the function returns vectors of different types? For example, return either `[1]` or `[1.0]` depending on the group.

Reply:
Added.

Reply:
That function always returns vectors of the same type. :-/

Reply:
does the line below this work? I thought I had it in that last commit...

Reply:
Well, no, it always return a value of the same type. You need something like `:b[1] == 1 ? [1] : [1.0]`.

Reply:
...or better `:b[1] == 1 ? [0.0] : [:b[1]]`, which is more likely to catch bugs since values change. BTW, the test would be better if there were more than two groups and values were not sorted.

Reply:
You are totally right. This behavior we need to pin down. 

If the first result is a scalar, then any 1-length vector down the line works. But if the first result if a vector of length one, it will throw an error that the length type doesnt match the group's length. 

We could do this vector-scalar check in every iteration of the loop instead of just the first one. 

```
d = DataFrame(n = 1:20, x = [3, 3, 3, 3, 1, 1, 1, 2, 1, 1, 2, 1, 1, 2, 2, 2, 3, 1, 1, 2])
g = groupby(d, :x)
@transform(g, t = :n[1] == 1 ? [0], [0.0]) # error because we check length for vector outputs
@transform(g, t = :n[1] == 1 ? 0 :  [0.0]) # no error
```

Reply:
Yes, you can copy the error message I throw in that case at https://github.com/JuliaData/DataFrames.jl/pull/1520.

But that's not what I meant originally, I just got the length of the vectors wrong. So we need tests both for cases where the lengths don't match, and for cases where lengths match but types change (the latter shouldn't throw an error). BTW, to cover all code paths you should also check that an error is thrown when the first vector doesn't have the right length, and when the first vector has the right length but the second doesn't.

Finally, as noted in an old comment, it would be nice to test returning a `CategoricalArray` and a `CategoricalValue`/`CategoricalString` (I know, writing tests is the worst part of development...).